modules/cudaarithm/src/cuda/absdiff_scalar.cu

```diff
@@ -78,33 +78,57 @@ namespace
     };
 
     template <typename SrcType, typename ScalarDepth>
-    void absDiffScalarImpl(const GpuMat& src, double value, GpuMat& dst, Stream& stream)
+    void absDiffScalarImpl(const GpuMat& src, cv::Scalar value, GpuMat& dst, Stream& stream)
     {
-        AbsDiffScalarOp<SrcType, ScalarDepth> op;
-        op.val = static_cast<ScalarDepth>(value);
+        typedef typename MakeVec<ScalarDepth, VecTraits<SrcType>::cn>::type ScalarType;
+
+        cv::Scalar_<ScalarDepth> value_ = value;
+
+        AbsDiffScalarOp<SrcType, ScalarType> op;
+        op.val = VecTraits<ScalarType>::make(value_.val);
         gridTransformUnary_< TransformPolicy<ScalarDepth> >(globPtr<SrcType>(src), globPtr<SrcType>(dst), op, stream);
     }
 }
 
 void absDiffScalar(const GpuMat& src, cv::Scalar val, bool, GpuMat& dst, const GpuMat&, double, Stream& stream, int)
 {
-    typedef void (*func_t)(const GpuMat& src, double val, GpuMat& dst, Stream& stream);
-    static const func_t funcs[] =
+    typedef void (*func_t)(const GpuMat& src, cv::Scalar val, GpuMat& dst, Stream& stream);
+    static const func_t funcs[7][4] =
     {
-        absDiffScalarImpl<uchar, float>,
-        absDiffScalarImpl<schar, float>,
-        absDiffScalarImpl<ushort, float>,
-        absDiffScalarImpl<short, float>,
-        absDiffScalarImpl<int, float>,
-        absDiffScalarImpl<float, float>,
-        absDiffScalarImpl<double, double>
+        {
+            absDiffScalarImpl<uchar, double>, absDiffScalarImpl<uchar2, double>, absDiffScalarImpl<uchar3, double>, absDiffScalarImpl<uchar4, double>
```

Comment:
I have integrated corrections matching your remarks.
About testing, I do not know (yet) how to integrate that properly in opencv tests.
For now, I use a local external program that runs :

`int depths[] = {CV_8U, CV_8S, CV_16U, CV_16S, CV_32S, CV_32F, CV_64F};
  int channels[] = {1, 3, 4};
  if (isGpuEnabled)
  for(int i = 0 ; i<sizeof(depths)/sizeof(int) ; ++i)
  {
    for(int j = 0 ; j<sizeof(channels)/sizeof(int) ; ++j)
    {
      int type = CV_MAKETYPE(depths[i], channels[j]);
      try{
        cv::cuda::GpuMat gpuMat1(cv::Size(512, 512), type);
        cv::cuda::GpuMat gpuMat2(cv::Size(512, 512), type);
        cv::Mat mat1;
        cv::Mat mat2;

        gpuMat1.setTo(cv::Scalar(64, 128, 192, 256));
        gpuMat2.setTo(cv::Scalar(64, 128, 192, 256));

        cv::cuda::subtract(gpuMat1, cv::Scalar(64, -64, 32, -1024), gpuMat1);
        cv::cuda::abs(gpuMat1, gpuMat1);
        gpuMat1.download(mat1);

        cv::cuda::absdiff(gpuMat2, cv::Scalar(64, -64, 32, -1024), gpuMat2);
        gpuMat2.download(mat2);

        cv::Mat diff = (mat1 != mat2);
        std::vector<cv::Mat> channels(diff.channels());
        cv::split(diff, channels);
        size_t countNonZero = 0;
        for(std::vector<cv::Mat>::const_iterator it = channels.begin() ; it != channels.end() ; ++it)
          countNonZero += cv::countNonZero(*it);
        if (!countNonZero)
          printf("equality for type %d\n", type);
        else
          printf("NOT equality for type %d\n", type);
      }
      catch(std::exception& e){
        printf("type %d, exception <%s>\n", type, e.what());
      }
    }
  }`